taps_to_cases/runner/src/main/java/com/microsoft/hydralab/t2c/runner/controller/IOSDriverController.java

```diff
@@ -0,0 +1,162 @@
+import com.microsoft.hydralab.t2c.runner.T2CAppiumUtils;
+import io.appium.java_client.appmanagement.ApplicationState;
+import io.appium.java_client.ios.IOSDriver;
+import org.openqa.selenium.Dimension;
+import org.openqa.selenium.JavascriptExecutor;
+import org.openqa.selenium.Point;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.interactions.PointerInput;
+import org.openqa.selenium.interactions.Sequence;
+import org.slf4j.Logger;
+
+import java.time.Duration;
+import java.util.List;
+
+public class IOSDriverController extends BaseDriverController {
+    private final IOSDriver iosDriver;
+    private String clipboardString;
+
+    public IOSDriverController(IOSDriver iosDriver, String udid, Logger logger) {
+        super(iosDriver, udid, logger);
+        this.iosDriver = iosDriver;
+    }
+
+    @Override
+    public void activateApp(String appPackageName) {
+        if (iosDriver.isAppInstalled(appPackageName)) {
+            if (iosDriver.queryAppState(appPackageName) != ApplicationState.RUNNING_IN_FOREGROUND) {
+                iosDriver.activateApp(appPackageName);
+            }
+        } else {
+            throw new RuntimeException("the app " + appPackageName + " is not installed");
+        }
+    }
+
+    @Override
+    public void terminateApp(String appPackageName) {
+        if (iosDriver.queryAppState(appPackageName) != ApplicationState.NOT_RUNNING &&
+                iosDriver.queryAppState(appPackageName) != ApplicationState.NOT_INSTALLED) {
+            iosDriver.terminateApp(appPackageName);
+        }
+    }
+
+    @Override
+    public void scroll(WebElement webElement, int xVector, int yVector) {
+        Point location = webElement.getLocation();
+        Dimension dimension = webElement.getSize();
+        int x = location.getX();
+        int y = location.getY();
+        int width = dimension.getWidth();
+        int height = dimension.getHeight();
+        int centerX = x + width / 2;
+        int centerY = y + height / 2;
+        PointerInput finger = new PointerInput(PointerInput.Kind.TOUCH, "finger");
+        Sequence dragNDrop = new Sequence(finger, 1);
+        dragNDrop.addAction(finger.createPointerMove(Duration.ofMillis(0),
+                PointerInput.Origin.viewport(), centerX, centerY));
+        dragNDrop.addAction(finger.createPointerDown(PointerInput.MouseButton.LEFT.asArg()));
+        dragNDrop.addAction(finger.createPointerMove(Duration.ofMillis(700),
+                PointerInput.Origin.viewport(), centerX + xVector, centerY + yVector));
+        dragNDrop.addAction(finger.createPointerUp(PointerInput.MouseButton.LEFT.asArg()));
+        iosDriver.perform(List.of(dragNDrop));
+
+        logger.info("centerX" + centerX + "centerY" + centerY);
+    }
+
+    @Override
+    public void swipe(String direction) {
+        //androidDriver.executeScript("mobile: scroll", ImmutableMap.of("direction", direction));
```

Comment:
This comment line should probably be removed.

Reply:
revised